File: src/main/java/com/bedirhanatasoy/game/tradingcard/game/Game.java
package com.bedirhanatasoy.game.tradingcard.game;

import com.bedirhanatasoy.game.tradingcard.player.Card;
import com.bedirhanatasoy.game.tradingcard.player.Deck;
import com.bedirhanatasoy.game.tradingcard.player.Player;
import com.bedirhanatasoy.game.tradingcard.rule.RuleEngine;
import com.bedirhanatasoy.game.tradingcard.rule.RuleType;

import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.Scanner;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Game {

    private static final int INITIAL_HEALTH = 30;

    private static final int INITIAL_MANA = 0;

    private static final int INITIAL_CARD_COUNT = 3;

    private static final List<Integer> INITIAL_DECK_MANA_COSTS =
            Arrays.asList(0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 8);

    private RuleEngine ruleEngine = new RuleEngine(this);

    private List<Player> players;

    private int activePlayerIndex = new Random().nextInt(2);

    private Player activePlayer;

    private GameState state = GameState.PLAYING;

    public Game() {
        players = IntStream.range(0, 2)
                .mapToObj(it -> {
                    System.out.println("Type a name for player " + (it + 1));
                    String playerName = getPlayerInput();
                    Deck deck = initializeDeck();
                    Player player = new Player(playerName, INITIAL_HEALTH, INITIAL_MANA, deck);
                    player.drawRandomCardsFromDeck(INITIAL_CARD_COUNT);
                    return player;
                }).collect(Collectors.toList());
        activePlayer = players.get(activePlayerIndex);
    }

    private Deck initializeDeck() {
        List<Card> cards = INITIAL_DECK_MANA_COSTS.stream()
                .map(Card::new)
                .collect(Collectors.toList());
        return new Deck(cards);
    }

    public void roundLoop() {
        int opponentPlayerIndex = (activePlayerIndex + 1) % 2;
        Player opponentPlayer = players.get(opponentPlayerIndex);

        ruleEngine.executeRules(RuleType.BEFORE_ROUND);
        System.out.println("\nRound started for " + activePlayer.getName());
        ruleEngine.executeRules(RuleType.ON_ROUND);
        cardSelectionLoop(activePlayer, opponentPlayer);
        activePlayerIndex = opponentPlayerIndex;
        activePlayer = players.get(activePlayerIndex);
        ruleEngine.executeRules(RuleType.AFTER_ROUND);
    }

    private void cardSelectionLoop(Player player, Player opponentPlayer) {
        showPlayersInfos();
        int selectedCardIndex;
        try {
            selectedCardIndex = askCardIndex(player);
        } catch (NumberFormatException e) {
            selectedCardIndex = -1;
        }

        while (selectedCardIndex >= 0 && selectedCardIndex < player.getCards().size()) {
            Card card = player.useCard(selectedCardIndex);
            if (card == null) {
                System.out.println(player.getName() + ", you don't have enough mana for this card! Please select another card");
            } else {
                opponentPlayer.decreaseHealth(card.getManaCost());
                System.out.println(player.getName() + " damaged [" + card.getManaCost() + "] to " + opponentPlayer.getName());
            }
            showPlayersInfos();
            selectedCardIndex = askCardIndex(player);
        }
    }

    private void showPlayersInfos() {
        System.out.println("\nPlayer Infos:");
        players.forEach(player -> System.out.println(player.getName()
                + " - Health: " + player.getHealth()
                + " - Mana: " + player.getMana()
                + " - Mana Slot: " + player.getManaSlot()
                + " - Cards: " + player.getCards().size()
                + " - Deck Cards: " + player.getDeck().getCards().size()));
    }

    private void showPlayerCards(Player player) {
        System.out.println("\n" + player.getName() + ", your current cards are:");
        List<Card> cards = player.getCards();
        IntStream.range(0, cards.size())
                .forEach(index -> System.out.println("[" + (index + 1) + "] => " + cards.get(index).getManaCost()));
    }

    private int askCardIndex(Player player) {
        showPlayerCards(player);
        System.out.println("\n" + player.getName() + ", type the index of the card to use or type `0` to skip this round");
        return Integer.valueOf(getPlayerInput()) - 1;
    }

    public void finishGame() {
        Player winner = players.stream().filter(player -> player.getHealth() > 0).findFirst().get();
        System.out.println("The winner is " + winner.getName() + "!!!");
        state = GameState.TERMINATED;
    }

    private String getPlayerInput() {
        return new Scanner(System.in).next();
    }


    public List<Player> getPlayers() {
        return players;
    }

    public Player getActivePlayer() {
        return activePlayer;
    }

    public GameState getState() {
        return state;
    }

}

File: src/main/kotlin/com/bedirhanatasoy/game/card/game/Game.kt
package com.bedirhanatasoy.game.card.game

import com.bedirhanatasoy.game.card.player.Card
import com.bedirhanatasoy.game.card.player.Deck
import com.bedirhanatasoy.game.card.player.Player
import com.bedirhanatasoy.game.card.rule.RuleEngine
import com.bedirhanatasoy.game.card.rule.RuleType
import java.util.*
import kotlin.random.Random

class Game {

    private val initialHealth = 30
    private val initialMana = 0
    private val initialCardCount = 3
    private val initialDeckManaCosts = listOf(0, 0, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 7, 8)

    private val ruleEngine = RuleEngine(this)
    private var activePlayerIndex = Random.nextInt(0, 1)
    val players: List<Player>
    var activePlayer: Player
    var state: GameState = GameState.PLAYING


    init {
        players = IntRange(0, 1).map {
            print("Type a name for player $it: ")
            val playerName = getPlayerInput()
            val deck = initialDeckManaCosts
                .map { Card(it) }
                .toMutableList()
                .let { Deck(it) }
            Player(playerName, initialHealth, initialMana, deck)
                .also { it.drawRandomCardsFromDeck(initialCardCount) }
        }

        activePlayer = players[activePlayerIndex]
    }

    fun roundLoop() {
        val opponentPlayerIndex = (activePlayerIndex + 1) % 2
        val opponentPlayer = players[opponentPlayerIndex]

        ruleEngine.executeRules(RuleType.BEFORE_ROUND)
        println("\nRound started for ${activePlayer.name}")
        ruleEngine.executeRules(RuleType.ON_ROUND)
        startCardSelection(activePlayer, opponentPlayer)
        activePlayerIndex = opponentPlayerIndex
        activePlayer = players[activePlayerIndex]
        ruleEngine.executeRules(RuleType.AFTER_ROUND)
    }

    fun startCardSelection(player: Player, opponentPlayer: Player) {
        showPlayerInfos()
        var selectedCardIndex = askCardIndexFromPlayer(player)
        while (selectedCardIndex >= 0 && selectedCardIndex < player.cards.size) {
            player.useCard(selectedCardIndex)
                ?.also { opponentPlayer.decreaseHealth(it.manaCost) }
                ?.also { println("${player.name} damaged [${it.manaCost}] to ${opponentPlayer.name}") }
                ?: println("${player.name}, you don't have enough mana for this card! Please select another card")
            showPlayerInfos()
            selectedCardIndex = askCardIndexFromPlayer(player)
        }
    }

    fun finishGame() {
        players.first { it.health > 0 }
            .also { println("The winner is ${it.name}!!!") }
            .also { state = GameState.TERMINATED }
    }

    fun showPlayerInfos() {
        println("\nPlayer Infos")
        players.forEach { println("${it.name} - Health: ${it.health} - Mana: ${it.mana} - Cards: ${it.cards.size} - Deck Cards: ${it.deck.cards.size}") }
    }

    private fun showPlayerCards(player: Player) {
        println("\n${player.name}, your current cards are:")
        player.cards.forEachIndexed { index, card ->
            println("[${index + 1}] => ${card.manaCost}")
        }
    }

    private fun askCardIndexFromPlayer(player: Player): Int {
        showPlayerCards(player)
        println("\n${player.name}, type the index of the card to use or type `0` to skip this round")
        return getPlayerInput().toIntOrNull()?.minus(1) ?: 0
    }


    private fun getPlayerInput(): String = Scanner(System.`in`).next()

}